src/base/Gas.ml

```diff
@@ -175,204 +172,235 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+        @@ GasCharge.SumOf
+             (GasCharge.SizeOf (GI.get_id s1), GasCharge.SizeOf (GI.get_id s2))
+    | Builtin_substr, [ s; i1; i2 ] ->
         pure
-        @@ Int.min (String.length s)
-             (Stdint.Uint32.to_int i1 + Stdint.Uint32.to_int i2)
-           * base
-    | Builtin_strlen, [ StringLit s ] -> pure @@ (String.length s * base)
-    | Builtin_to_string, [ l ] ->
-        let%bind c = literal_cost l in
-        pure @@ (c * base)
+        @@ GasCharge.MinOf
+             ( GasCharge.SizeOf (GI.get_id s),
+               GasCharge.SumOf
+                 ( GasCharge.ValueOf (GI.get_id i1),
+                   GasCharge.ValueOf (GI.get_id i2) ) )
+    | Builtin_strlen, [ s ] -> pure @@ GasCharge.SizeOf (GI.get_id s)
+    | Builtin_to_string, [ l ] -> pure @@ GasCharge.SizeOf (GI.get_id l)
     | _ -> fail0 @@ "Gas cost error for string built-in"
 
-  let crypto_coster op args base =
-    let%bind types = mapM args ~f:literal_type in
-    let div_ceil x y = if x % y = 0 then x / y else (x / y) + 1 in
+  let crypto_coster op args types =
     match (op, types, args) with
-    | Builtin_eq, [ PrimType Bystr_typ; PrimType Bystr_typ ], [ ByStr a1; _ ] ->
-        pure (Bystr.width a1 * base)
+    | Builtin_eq, [ PrimType Bystr_typ; PrimType Bystr_typ ], [ a1; _ ] ->
+        pure @@ GasCharge.SizeOf (GI.get_id a1)
     | Builtin_eq, [ a1; a2 ], _
       when is_bystrx_type a1 && is_bystrx_type a2
            && Option.(value_exn (bystrx_width a1) = value_exn (bystrx_width a2))
       ->
-        pure @@ (Option.value_exn (bystrx_width a1) * base)
+        let width = Option.value_exn (bystrx_width a1) in
+        pure @@ GasCharge.StaticCost width
     | Builtin_to_uint256, [ a ], _
       when is_bystrx_type a && Option.value_exn (bystrx_width a) <= 32 ->
-        pure (32 * base)
+        pure @@ GasCharge.StaticCost 32
     | Builtin_sha256hash, _, [ a ] | Builtin_schnorr_get_address, _, [ a ] ->
         (* Block size of sha256hash is 512 *)
-        pure @@ (div_ceil (String.length (pp_literal a)) 64 * 15 * base)
+        let s = GasCharge.SizeOf (GI.get_id a) in
+        let n = GasCharge.StaticCost (64 * 15) in
+        pure (GasCharge.DivCeil (s, n))
     | Builtin_keccak256hash, _, [ a ] ->
         (* Block size of keccak256hash is 1088 *)
-        pure @@ (div_ceil (String.length (pp_literal a)) 136 * 15 * base)
+        let s = GasCharge.SizeOf (GI.get_id a) in
+        let n = GasCharge.StaticCost (136 * 15) in
+        pure (GasCharge.DivCeil (s, n))
     | Builtin_ripemd160hash, _, [ a ] ->
         (* Block size of ripemd160hash is 512 *)
-        pure @@ (div_ceil (String.length (pp_literal a)) 64 * 10 * base)
-    | Builtin_schnorr_verify, _, [ _; ByStr s; _ ]
-    | Builtin_ecdsa_verify, _, [ _; ByStr s; _ ] ->
-        let x = div_ceil (Bystr.width s + 66) 64 in
-        pure @@ ((250 + (15 * x)) * base)
+        let s = GasCharge.SizeOf (GI.get_id a) in
+        let n = GasCharge.StaticCost (64 * 15) in
```

Comment:
The previous gas charge was size * 64 * 10, but now it's size * 64 * 15. I don't know which one is correct.